### ISSUE TITLE
Query results contain some deleted ids

### ISSUE THREAD

**Issue by NotSky** [milvus-io/milvus#4897]
1:通过调用deleteEntityByID已经删掉了对应的向量。
2:通过getEntitybyID查询时查不到。
3:但是通过search搜相似时，居然又查出来了。

麻烦研发给看看呀。
512维向量数总共100w。

场景是：实时写入，实时删除，实时查询。

**Comment by NotSky** [comment]
@shengjun1985 这个问题给看看呢

**Comment by shengjun1985** [comment]
Excuse me, did you test on a single computer or on multiple computers with Mishard?
And what's your Milvus version?

**Comment by NotSky** [comment]
a single computer !
version:0.10.5

**Comment by yhmo** [comment]
@NotSky 
Could you provide a python script that can reproduce this problem?

**Comment by yhmo** [comment]
How many clients in your scenario?
Did you start two clients, one client insert and delete, the other search?

**Comment by NotSky** [comment]
12个client ，每个client都有可能insert或者delete，另外还有12个client进行search

**Comment by yhmo** [comment]
I could not reproduce the issue.
My steps:

1. start a milvus server by  docker image 'milvusdb/milvus:0.10.5-cpu-d010621-4eda95'
2. run the following script
 the following script do these things:
 - create a collection(dim=512)
 - insert 100M vectors into the colection, each vector is constructed by its id
    for example, for vector id = 5, its value is [5, 5, 5, ...... , 5],  for vector id = 10000, its value is [10000, 10000, ...... , 10000]
 - create 10 threads to delete entity
 - create 10 threads send to search at the same time
```
import time
import threading
from milvus import Milvus, IndexType, Status
from milvus.client.types import MetricType

SERVER_ADDR = "127.0.0.1"
SERVER_PORT = '19530'

COLLECTION_NAME = 'TEST'
COLLECTION_DIMENSION = 512

INDEX_TYPE = IndexType.IVF_FLAT
INDEX_PARAM = {'nlist': 1024}

SEARCH_PARAM = {'nprobe': 16}

MILVUS = Milvus(host=SERVER_ADDR, port=SERVER_PORT)


def create_collection():
    tb_param = {'collection_name': COLLECTION_NAME, 'dimension': COLLECTION_DIMENSION, 'index_file_size': 1024,
                'metric_type': MetricType.L2}
    MILVUS.create_collection(tb_param)
    MILVUS.create_index(COLLECTION_NAME, INDEX_TYPE, INDEX_PARAM)
    print(COLLECTION_NAME, 'created')

def drop_collection():
    MILVUS.drop_collection(collection_name=COLLECTION_NAME)
    print(COLLECTION_NAME, 'dropped')

def count_collection():
    status, row_count = MILVUS.count_entities(collection_name=COLLECTION_NAME)
    print("row count =", row_count)

def gen_vec_list(vec_ids):
    vec_list = []
    for id in vec_ids:
        vec = [id for k in range(COLLECTION_DIMENSION)]
        vec_list.append(vec)
    return vec_list

def insert():
    batch_count = 10000
    for i in range(100):
        print("Insert no.", i)
        vec_ids = [k + batch_count*i for k in range(batch_count)]
        vec_list = gen_vec_list(vec_ids)
        status, result = MILVUS.insert(collection_name=COLLECTION_NAME, records=vec_list, ids=vec_ids)
        # print(status)
    MILVUS.flush(collection_name_array=[COLLECTION_NAME])
    print("insert done")

def search(id):
    for k in range(5):
        print("search no.", k)
        time.sleep(0.5)
        vec_ids = [id]
        query_vectors = gen_vec_list(vec_ids)
        status, results = MILVUS.search(collection_name=COLLECTION_NAME, query_records=query_vectors,
                                        top_k=3, params=SEARCH_PARAM)
        print(results)

    print('search done')

def delete(step):
    for i in range(1000000):
        if i%step == 0:
            id_arr = [i]
            status = MILVUS.delete_entity_by_id(collection_name=COLLECTION_NAME, id_array=id_arr)
            # print(status)
            MILVUS.flush(collection_name_array=[COLLECTION_NAME])
            print("delete", i)
    print("delete done")

if __name__ == "__main__":
    drop_collection()
    create_collection()
    insert()
    count_collection()

    threads = []
    for i in range(10):
        step = (i+1)*10000
        x = threading.Thread(target=delete, args=(step,))
        threads.append(x)
        x.start()

    for i in range(10):
        id = i*1000
        x = threading.Thread(target=search, args=(id,))
        threads.append(x)
        x.start()

    for th in threads:
        th.join()

```

3. run the second script to get entity by id and search the deleted entities
   
```
import sys
from milvus import Milvus, IndexType, Status
from milvus.client.types import MetricType

SERVER_ADDR = "127.0.0.1"
SERVER_PORT = '19530'

COLLECTION_NAME = 'TEST'
COLLECTION_DIMENSION = 512

SEARCH_PARAM = {'nprobe': 16}

MILVUS = Milvus(host=SERVER_ADDR, port=SERVER_PORT)

def gen_vec_list(vec_ids):
    vec_list = []
    for id in vec_ids:
        vec = [id for k in range(COLLECTION_DIMENSION)]
        vec_list.append(vec)
    return vec_list

def get_entities(id):
    id_arr = [id]
    status, vectors = MILVUS.get_entity_by_id(collection_name=COLLECTION_NAME, ids=id_arr)
    print("get entity", id)
    one_vector = vectors[0]
    if len(one_vector) != 0:
        print("deleted entity still exist", id)
        sys.exit(2)
    else:
        print("doesn't exist")

    print("get_entities done")

def search(id):
    vec_ids = [id]
    query_vectors = gen_vec_list(vec_ids)
    print("search entity", id)
    status, results = MILVUS.search(collection_name=COLLECTION_NAME, query_records=query_vectors,
                                    top_k=3, params=SEARCH_PARAM)
    found = False
    for each_query in results:
        for each_score in each_query:
            print("id = ", each_score.id, " distance = ", each_score.distance)
            if each_score.id == id:
                found = True

    if found:
        print("deleted entity still can be search", id)
        sys.exit(2)
    print("search done")

if __name__ == "__main__":
    for i in range(100000):
        for k in range(10):
            if i%((k+1)*10000) == 0:
                get_entities(i)
                search(i)

```
If any deleted entity is still exist, the script will give an error: "deleted entity still can be search", but it didn't throw the error.

**Comment by NotSky** [comment]
modify search top_k=100 retest?

**Comment by yhmo** [comment]
No, top_k=100 also cannot reproduce it.
Looks like it happens under certain condition, hard to reproduce.

**Comment by shengjun1985** [comment]
@NotSky
Excuse me, have you upgraded Milvus to v1.1.0?
And does this issue still happen on v1.1.0?

**Comment by shengjun1985** [comment]
@NotSky
Did you use the CPU version or the GPU version when the bug occurred?

**Comment by shengjun1985** [comment]
I seem to find the bug by reviewing source code. 
In some extreme cases, the `bitset` and `bloom filter` in the cache will not be updated.
We will try to fix it as soon as possible.

**Comment by shengjun1985** [comment]
It should be fixed in branch 1.1
If you want to use it as soon as possible, you can checkout the branch and compile it locally.
Otherwise, please wait for the next release.

Reference:
[Install Milvus from Source Code](https://github.com/milvus-io/milvus/blob/1.1/INSTALL.md)
